### ISSUE TITLE
[Discussion] Reduce 1M Metal peak via comp_mask sizing and phase-separated DSpark residency

### ISSUE THREAD

**Issue by apetersson** [antirez/ds4#643]
> **Disclosure:** This draft/discussion was prepared and filed by **Codex (OpenAI)** on behalf of `@apetersson` while attempting to build a 128 GB-optimized, selectively native-MXFP4 DeepSeek V4 Flash quant on the experimental `ds4f-mxfp4` branch. The audited revision is `4893e0c40fba03dbc85555faeb035799aa04e0b6`. The memory figures below are source-derived allocation estimates, not yet measured peak-footprint results; please independently validate them.

## Discussion goal

Could DeepSeek V4 Flash's 1,048,576-token Metal peak be reduced without lowering quantization quality or globally shrinking the prefill batch?

Two potentially complementary opportunities appeared during the audit:

1. Right-size the large `comp_mask_by_tier` allocation.
2. Avoid overlapping the large prefill workspace with DSpark support-model residency.

This is intentionally a design discussion rather than a claim that either change is immediately safe.

## 128 GB profile context

The candidate package is approximately:

- Main GGUF: `95.7644 GiB`
- Native-down DSpark support: `6.7965 GiB`
- Combined weights: `102.5609 GiB`

On an M1 Ultra with Metal reporting an approximately `121.60 GiB` recommended working set, source-based estimates at `ctx=1,048,576`, prefill chunk `4096`, are:

- Persistent compressed KV: about `8.031 GiB`
- Raw KV: about `0.357 GiB`
- Full chunk-scaled graph workspace: about `11.959 GiB`
- Additional verifier/logit/host buffers: roughly `0.34 GiB`
- Total: about `123.24 GiB`

Reducing the global prefill chunk works (`2048` is about `117.10 GiB`; `1024` is about `114.02 GiB`), but it also reduces the batch used by unrelated prefill stages.

## Idea A: audit and right-size `comp_mask_by_tier`

The graph currently allocates both buffers as:

```c
g->indexer_scores_by_tier[t] = ds4_gpu_tensor_alloc_ptr_on(
    t, (uint64_t)g->comp_cap * pc * sizeof(float));
g->comp_mask_by_tier[t] = ds4_gpu_tensor_alloc_ptr_on(
    t, (uint64_t)g->comp_cap * pc * sizeof(float));
```

At 1M context, ratio-4 `comp_cap` is approximately 262,146. With `pc=4096`, each buffer is about 4 GiB.

`indexer_scores` appears to need the compressed-rows by batch shape for batch score/top-k. In contrast, the normal indexed batch attention path consumes compact selected indices; `comp_mask` appears to be used mainly by one-row/fallback attention and small top-value scratch paths.

Questions for maintainers:

- Is any Metal path expected to consume `comp_cap * pc` live mask elements?
- Could `comp_mask` be sized to the maximum actual fallback row count, or split from the small top-value scratch use?
- Are CUDA tensor-parallel uses the reason for the shared oversized allocation, and could backend-specific sizing make this safe?

If the full matrix is unnecessary, reclaiming close to 4 GiB could move the estimated full package from `123.24` to about `119.24 GiB` without changing weights or batch size. That would fit nominally, though the remaining margin would still require real peak testing.

## Idea B: phase-separate prefill workspace and DSpark residency

DSpark does not accelerate prefill, yet its approximately 6.8 GiB support model and verifier state coexist with the maximum prefill workspace.

A possible lifecycle would be:

1. Create/prefill the main-model session without mapping or requesting residency for the support GGUF.
2. Finish prefill and release or shrink the prefill-only workspace.
3. Map the DSpark support model and allocate verifier state for decode.
4. If a later prompt extension needs batch prefill, safely reverse the transition or use a bounded extension workspace.

This trades implementation complexity for a lower peak; it should not alter target-model arithmetic.

Questions:

- Can the support model be bound lazily after the main session's KV has been populated?
- Which prefill tensors must remain live for interactive continuation?
- Would a shared/lazily recreated workspace be preferable to retaining every prefill tensor?
- How should cancellation and server session reuse coordinate the transition?

## Suggested validation

- Add explicit allocation-bound assertions for every `comp_mask` consumer.
- Compare logits/decode consistency before and after any sizing change.
- Exercise zero-prefix prefill, resumed prefill, mixed attention, one-token decode and speculative verification.
- Test session extension after entering DSpark decode mode.
- Report mapped main/support weights, raw/compressed KV, indexer scratch, other prefill tensors, verifier state, runtime tensor peak and process physical footprint separately.
- Benchmark chunk `4096`, `2048` and `1024` on the same M1 Ultra.

If maintainers agree either direction is sound, the implementation should probably be split into small independent patches rather than one combined change.
